snapshots/windows/windows.go

```diff
@@ -273,23 +336,28 @@ func (s *snapshotter) Close() error {
 
 func (s *snapshotter) mounts(sn storage.Snapshot) []mount.Mount {
 	var (
-		roFlag           string
-		source           string
-		parentLayerPaths []string
+		roFlag string
 	)
 
-	if sn.Kind == snapshots.KindView {
-		roFlag = "ro"
-	} else {
+	if sn.Kind == snapshots.KindActive {
 		roFlag = "rw"
+	} else {
+		roFlag = "ro"
 	}
 
-	if len(sn.ParentIDs) == 0 || sn.Kind == snapshots.KindActive {
-		source = s.getSnapshotDir(sn.ID)
-		parentLayerPaths = s.parentIDsToParentPaths(sn.ParentIDs)
-	} else {
-		source = s.getSnapshotDir(sn.ParentIDs[0])
-		parentLayerPaths = s.parentIDsToParentPaths(sn.ParentIDs[1:])
+	source := s.getSnapshotDir(sn.ID)
+	parentLayerPaths := s.parentIDsToParentPaths(sn.ParentIDs)
+
+	mountType := "windows-layer"
+
+	if len(sn.ParentIDs) == 0 {
+		// If there are no parents, it's a bind-mount, since we cannot
+		// have a windows-layer mount with no parents.
+		// https://github.com/microsoft/hcsshim/issues/853
+		// The only time this is weird is if an Active snapshot has a
+		// diff applied to it, and is then mounted to a folder; in this
+		// case, the windows-layer data structures will be visible.
+		mountType = "bind"
```

Comment:
I discovered (https://github.com/microsoft/hcsshim/issues/853) that my assertion here is not quite correct. It turns out we probably _can_ create empty windows-layers, but it'll need a bit of extra work in hcsshim (or reproducing the code locally) to work on Windows 1809/LTSC 2019 and earlier.

Note: I haven't actually tested that this will produce a valid sandbox.vhdx that HCS will mount as a read-write layer. The [use-case in hcsshim](https://github.com/microsoft/hcsshim/blob/fc27c5026e6ff001dc1b171b99bda7bb3dcf6e78/test/cri-containerd/runpodsandbox_test.go#L1265-L1278) only mounts the so-created VHDX into a container, but doesn't mount it on the host filesystem, which is the use-case here.

Would that be a better direction? It would certainly make the whole system more-consistent by removing the "bind" code paths introduced in 8d7df6578077da72b3e3863a69ceac3bc1e1f40e, in favour of always creating and mounting snapshots as windows-layers, and it would ensure that _any_ snapshot can be committed and used as a parent to another snapshot, whether it came from an applied OCI layer, or by user action.

Effectively, this would mean 'FROM scratch` becomes feasible for Windows containers. Although the lack of an OS prevents it from being used as a container root FS, it would unblock, e.g., https://github.com/microsoft/hcsshim/issues/750.

Reply:
To surface the outcome in https://github.com/microsoft/hcsshim/issues/853: I couldn't get a valid parentless Windows Containers layer created though the APIs exposed in hcsshim. It's possible the RS5 computestorage.dll APIs visible in abandoned https://github.com/microsoft/hcsshim/pull/407 might support this, but I didn't try them, and would frankly just be poking around in the dark anyway as I'm unaware of any other documentation for those APIs.

Reply:
Based on more experiments in https://github.com/microsoft/hcsshim/issues/853, it turns out that we can create parentless WCOW layers, they are created _after_ the data is populated, so it's a bind-mount of _Files/_ in both user-created and imported-from-server parentless-layer cases, which is nice.